Add `Coinbase(Transaction)` newtype to distinguish coinbase transactions

jrakibi opened the pull request:
Coinbase transactions are structurally and semantically different from normal transactions.

We introduce a distinct `Coinbase(Transaction)` newtype to make this distinction explicit at the type level, helping to catch logic that shouldn't apply to coinbases (e.g attempting to look up coinbase inputs in the UTXO set).

This PR is split into several patches to keep changes reviewable and to document the rationale behind each step.

- Closes #4437 
- Closes #4543

tcharding commented:
Why is this draft mate? Are you chasing concept ack or full review?

jrakibi commented:
> Why is this draft mate? Are you chasing concept ack or full review?

I just marked it as a draft to double-check a few things before opening it up. it's ready for full review now

tcharding reviewed (commented):
I'm going to give an unhelpful review: something doesn't quite feel right about this but I can't work out what. Perhaps one of the other fellas can shed light on it. The fact that we map all the errors using `|_|` is a hint I think. I'll re-review after the error variant thing is fixed, maybe something will fall out.

jrakibi commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1130,6 +1132,46 @@ impl InputWeightPrediction {
+
+
+impl<'a> TryFrom<&'a Transaction> for &'a Coinbase {
+    type Error = CoinbaseError;
```
I feel like a better approach for error handling would be to rename `Bip34Error` to something more generic and use it for both methods: `coinbase()` and `bip34_block_height()`

That way I can have all error variants related to coinbase operations in one place instead of creating two separate error types !

tcharding replied:
In this repo we favour having exact errors that only include variants that are returned by the function.

Kixunil reviewed (dismissed):
At minimum, I want the `transparent_newtype` macro to be used so that we don't have to burn brain power on thinking about `unsafe`. The remaining requests are fairly important though might be controversial.

jrakibi commented:
Apologise for the delay in addressing your feedback. I think all review comments have been addressed:

- Used `transparent_newtype` macro for the `Coinbase` type, which eliminate `unsafe` concerns 
- Implemented block invariant validation - blocks must have at least one transaction, enforced during `Block::validate()` with `NoTransactions` and `InvalidCoinbase` error variants

- Added `assume_first_transaction` & `assume_first_transaction_ref` methods as suggested 

- Separated error concerns - moved coinbase validation to block-level validation rather than adding variants to `Bip34Error` that would never be returned by BIP34 parsing
- Made `coinbase()` infallible for checked blocks

jrakibi commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1130,6 +1132,46 @@ impl InputWeightPrediction {
+
+    fn try_from(tx: &'a Transaction) -> Result<Self, Self::Error> {
+        if tx.is_coinbase() {
+            Ok(unsafe { &*(tx as *const Transaction as *const Coinbase) })
```
Done! Used `transparent_newtype` macro and removed the `try_from` approach

apoelstra commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1130,6 +1132,46 @@ impl InputWeightPrediction {
+
+impl Coinbase {
+    /// Computes the [`Txid`] of this coinbase transaction.
+    pub fn compute_txid(&self) -> Txid {
```
In 6dce3540536d1e6439f6a576785053b501eaa7c5:

Why have these methods? They are available through `Deref`.

jrakibi replied:
I've moved away from using implicit inheritance via `Deref` to avoid inheriting the entire Transaction API. Instead, we now expose only the methods that make sense for coinbase transactions.

`inner()` serves as an explicit escape hatch when full Transaction access is needed. This forces callers to be intentional about when they need the underlying Transaction functionalities.

apoelstra commented on primitives/src/block.rs:
```diff
@@ -76,6 +76,12 @@ impl Block<Unchecked> {
 
+    /// Adds a transaction to the block.
+    #[inline]
+    pub fn push_transaction(&mut self, transaction: Transaction) {
```
This method isn't used in this PR. Presumably you want to drop it, since adding it properly would be a bit involved (I guess we want to maintain at least the rightmost merkle path in the block?)

Kixunil replied:
I think it fits better into `block::Builder` where the merkle root is only computed once all transactions were added.

jrakibi commented on bitcoin/src/blockdata/block.rs:
```diff
@@ -505,6 +505,42 @@ impl std::error::Error for ValidationError {
+
+impl fmt::Display for CoinbaseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use CoinbaseError::*;
```
Noted. Also, I'm not using `CoinbaseError` anymore, I’ve already removed it

tcharding commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1126,6 +1126,52 @@ impl InputWeightPrediction {
+    /// Computes the [`Wtxid`] of this coinbase transaction.
+    pub fn compute_wtxid(&self) -> Wtxid {
+        self.0.compute_wtxid()
+    }
```
API suggestion:

```rust
    /// Returns a reference to the inner transaction.
    pub fn as_inner(&self) -> &Transaction { &self.0 }

    /// Returns the inner transaction.
    pub fn into_inner(self) -> Transaction { self.0 }
```

ref: https://rust-lang.github.io/api-guidelines/naming.html

jrakibi replied:
Good idea, added `as_inner` and `into_inner` as suggested
thanks for pointing to the guidelines

tcharding commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1126,6 +1126,52 @@ impl InputWeightPrediction {
+
+    impl Coinbase {
+        /// Creates a reference to `Coinbase` from a reference to the inner `Transaction`.
+        pub fn from_ref(inner: &_) -> &Self;
```
From looking at other call sites I think this should probably be `from_transaction_ref`.

tcharding replied:
Or if there is not supposed to be an invariant perhaps it should be `from_first_transaction_ref` or something (to match the naming of the other constructors).

apoelstra replied:
I think it should be public. People will create coinbase transactions in all sorts of ways, including by reading them off the wire in ad-hoc mining protocols.

jrakibi replied:
Renamed the method to `assume_coinbase_ref` to reflect that the caller must ensure it's a valid coinbase. 

I Kept it public as per the feedback on ad-hoc mining protocols. Open to further changes if needed!

tcharding commented:
Can you move the last commit into the commit that originally creates the function please mate.

This is repository specific, other repos may, and will, have different strategies.   In this repo after review we make changes to the patch set and then force push. Its a reasonably common strategy, you should be able to find someone online explaining it. Or this prompt into opus gave me a pretty good response "when working on github with git why do some projects fix patches and force push instead of adding new patches onto the end after review".

tcharding commented on bitcoin/src/blockdata/block.rs:
```diff
@@ -293,8 +305,13 @@ impl BlockCheckedExt for Block<Checked> {
         size
     }
 
-    /// Returns the coinbase transaction, if one is present.
-    fn coinbase(&self) -> Option<&Transaction> { self.transactions().first() }
+    /// Returns the coinbase transaction.
+    ///
+    /// This method is infallible because validation ensures a valid coinbase transaction is always present.
+    fn coinbase(&self) -> &Coinbase {
```
nit: On the trait method impls we don't need to repeat the docs. If you are keen you could add a patch at the front that removes the original line also the one on `bip34_block_height`.

tcharding replied:
In ba5fb268fdf25281aeef8ab462bde506a00229c2

jrakibi replied:
Done

tcharding reviewed (approved):
ACK 23ef6c4b73b6187920ada826837459c0f0a415b8

Kixunil commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1141,6 +1141,41 @@ internals::transparent_newtype! {
+    }
+
+    /// Computes the [`Wtxid`] of this coinbase transaction.
+    pub fn compute_wtxid(&self) -> Wtxid {
```
IIRC coinbase transactions have a constant `Wtxid` so this method looks misleading and probably shouldn't exist.

Kixunil reviewed (changes requested):
At minimum, the compile error needs to be fixed.

apoelstra commented on bitcoin/src/blockdata/block.rs:
```diff
@@ -117,6 +118,14 @@ impl BlockUncheckedExt for Block<Unchecked> {
+            return Err(InvalidBlockError::NoTransactions);
+        }
+
+        if !transactions[0].is_coinbase() {
```
I think `is_coinbase` is fine. Though I think we should have an explicit conversion method that does `Transaction` -> `Coinbase` and this should be redefined just as `self.to_coinbase().is_ok()`.

In principle any coinbase-looking transaction "is a coinbase" in the sense that you could make a block with it as the first transaction.

Kixunil replied:
> In principle any coinbase-looking transaction "is a coinbase" in the sense that you could make a block with it as the first transaction.

Yes, but I think the `Coinbase` type should also encode that it actually *is* the first, not just that it *could* be since it saves a check and a bunch of disclaimers all over the API saying "if this is not actually the first transaction the behavior is unspecified".

apoelstra replied:
But it actually *is* the first transaction of any block that has it as the first transaction. I don't see how we can encode in our API some requirement that such a block actually be computed and propagated (or whatever the criteria for "is" is).

The name `looks_like_coinbase` is confusing and hard to find, and I don't think it's any more correct.

Kixunil replied:
> that such a block actually be computed and propagated

That'd be too much and I didn't propose that. What I'm proposing is that either:

* You obtained it from a `Block` (or manually parsing a block) and it was the first one there.
* You constructed it with the intention to insert it as the first transaction in a block (manually or through `Block`, thoutg I'd prefer a builder there).

Anyway, if you're still unconvinced then I want this at least documented.

apoelstra commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1141,6 +1141,41 @@ internals::transparent_newtype! {
+        Self::assume_coinbase_ref(tx)
+    }
+
+    /// Returns a reference to the inner transaction.
```
How could it? A coinbase is a transaction.

Kixunil replied:
E.g. trying to fetch "previous output" in a UTXO set.

apoelstra replied:
Ah, that's a good point. Ok, I'd be happy with such a comment. We should clarify that "nonsensical usage" means "using the prevout as though it were a real prevout".

jrakibi replied:
Great point

Added:
```rust
/// Warning: The coinbase input contains a dummy prevout that should not be treated as a real prevout
```

Kixunil replied:
IIRC, all prevouts are dummy, not just one, so it likely should be "dummy prevouts".

jrakibi commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1141,6 +1141,41 @@ internals::transparent_newtype! {
+    }
+
+    /// Returns a reference to the inner transaction.
+    pub fn as_inner(&self) -> &Transaction { &self.0 }
```
I’m assuming we’re moving away from using `into_*`  also, so I renamed both methods and added #[doc(alias = "...")]

jrakibi commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1141,6 +1141,41 @@ internals::transparent_newtype! {
+    /// The caller must ensure that this transaction is indeed the first transaction in a valid block.
+    pub fn assume_first_transaction_ref(tx: &Transaction) -> &Self {
+        Self::assume_coinbase_ref(tx)
+    }
```
Done. Removed the method for now

jrakibi commented on bitcoin/src/blockdata/block.rs:
```diff
@@ -328,8 +328,8 @@ impl BlockCheckedExt for Block<Checked> {
-        let cb = self.coinbase().ok_or(Bip34Error::NotPresent)?;
-        let input = cb.input.first().ok_or(Bip34Error::NotPresent)?;
+        let cb = self.coinbase();
+        let input = cb.as_inner().input.first().ok_or(Bip34Error::NotPresent)?;
```
All commits should now compile successfully

apoelstra commented:
In 1ff6e943ae60e04fce902dccd820c49582b6e6fb:

Compilation fails.

```
error[E0599]: no method named `unwrap` found for reference `&Coinbase` in the current scope
   --> src/blockdata/block.rs:548:37
    |
548 |         assert_eq!(block.coinbase().unwrap().compute_txid().to_string(), cb_txid);
    |                                     ^^^^^^ method not found in `&Coinbase`

error: aborting due to previous error
```

tcharding commented:
@jrakibi FYI Andrew tests each commit but in CI we only check the final state, hence why he posted the commit ID that is broken for you. Said another way, we have policy that every commit has to pass tests even though the cloud CI set up does not enforce this.

apoelstra reviewed (approved):
ACK a67151a81d061894c63b4d38774c5e52127c8b49; successfully ran local tests

Kixunil commented on bitcoin/src/blockdata/block.rs:
```diff
@@ -311,8 +317,8 @@ impl BlockCheckedExt for Block<Checked> {
-        let cb = self.coinbase().ok_or(Bip34Error::NotPresent)?;
-        let input = cb.input.first().ok_or(Bip34Error::NotPresent)?;
+        let cb = self.coinbase();
+        let input = cb.as_transaction().input.first().ok_or(Bip34Error::NotPresent)?;
```
This needs to be infallible since a valid coinbase is guaranteed to have at least one input.

tcharding reviewed (approved):
ACK 394fa537508674388f2cd4cdb23eccbc51d599ae

apoelstra reviewed (approved):
ACK 394fa537508674388f2cd4cdb23eccbc51d599ae; successfully ran local tests

apoelstra commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1169,6 +1169,43 @@ internals::transparent_newtype! {
+    /// 
+    /// For coinbase transactions, this is always `Wtxid::COINBASE`.
+    #[doc(alias = "compute_wtxid")]
+    pub fn wtxid(&self) -> Wtxid {
```
No, I think what you've done here is the best strategy.
